### DISCUSSION TITLE
Refactor

### DISCUSSION THREAD

**Discussion by KarolDenst** [Non-Euclidean-World/Hyper#29] [General]
Scene.cs is a mess and everything is in the same project. Here is an idea of how we could solve both of these issues.

Treat Scene.cs like a database holding our chunks, characters... For each type of body (characters, projectiles, chinks...) create a service class and a controller class. The controller will just have a Render() method and will just call service for other things (like in web applications). The service will have the logic. For each type of body we create a new project and move all associated classes there and create controller + service. We keep things like Texture.cs and Shader.cs in a new project Hyper.Common. This way each part of our application will be in its own separate project and it will be very easy to communicate with said part (just call the appropriate method in controller). 

This way it will be clear where to put what and we won’t have inconsistencies. You just separate the logic like you would separate the logic in a classic web API.

**Comment by AleksyBalazinski** [comment]
As of now, I have two concerns:
- I'm not entirely convinced if the ideas from web dev translate intuitively into our problem. These abstractions definitely lead to a more structured code, that's granted, but I just don't know if they aren't too loosely coupled to the domain (I really, really believe that DDD should be the primary decision driver)
- I don't know if that scales well: adding new type of body introduces a new project.

I'll definitely give it more thought though

**Reply by KarolDenst** [reply]
I don't understand this part:
loosely coupled to the domain 

As for scaling how many types of bodies could there be?

**Comment by KarolDenst** [comment]
Here is how i see it:
![Architecture drawio](https://github.com/Non-Euclidean-World/Hyper/assets/95054664/78802d64-718f-45c0-bbbf-a87d9a559f58)
